[TBDGen] Fix a deserialization crash in TBDGen

zixu-w opened the pull request:
Fix a deserialization failure crash in `TBDGenVisitor`, where an
internal protocol conformance with a project internal symbol witness for
an `associatedtype` triggers a crash inside `addConformances` because
the module containing the project internal symbol is not available.

Fixes rdar://93983322

zixu-w committed:
[TBDGen] Fix a deserialization crash in TBDGen

Fix a deserialization failure crash in `TBDGenVisitor`, where an
internal protocol conformance with a project internal symbol witness for
an `associatedtype` triggers a crash inside `addConformances` because
the module containing the project internal symbol is not available.

Fixes rdar://93983322

zixu-w commented:
@swift-ci Please test

xymus commented on lib/TBDGen/TBDGen.cpp:
```diff
@@ -479,6 +479,10 @@ void TBDGenVisitor::addConformances(const IterableDeclContext *IDC) {
                             ConformanceLookupKind::NonInherited)) {
     auto protocol = conformance->getProtocol();
+    if (Opts.PublicSymbolsOnly &&
+        getDeclLinkage(protocol) != FormalLinkage::PublicUnique)
```
Do we need to also accept `FormalLinkage::PublicNonUnique` to to cover protocols imported from Objective-C?

zixu-w replied:
For this particular workaround, I don't think Objective-C protocols are relevant because the crash occurs specifically with internal protocols with associated types.
And other places inside the `TBDGenVisitor` also only accept `PublicUnique` so I'm trying to keep it concise and consistent.
We can add `PublicNonUnique` later if it's causing problems.

zixu-w committed:
[TBDGen] Simplify test case TBD/rdar93983322.swift

Framework model is not related to the issue. Remove framework layouts
and search paths.

zixu-w commented:
@swift-ci Please smoke test macOS
